### DISCUSSION TITLE
How to add custom UCI-compatible engines

### DISCUSSION THREAD

**Discussion by C-Advait** [franciscoBSalgueiro/en-croissant#299] [Q&A]
Hi community,

I am working on trying to enable/use a _custom_ UCI-compatible engine [sunfish](https://github.com/thomasahle/sunfish) with the en-croissant GUI. The main issue is that once I add the engine binary to en-croissant, I am unable to play against it. It seems that the engine does not return a move. However I can run this binary through a terminal and communicate via UCI commands to get valid responses. 
I am trying to understand where and why there appears to be a communication breakdown between the engine and the en-croissant interface.
I have detailed my workflow below.

## Installing the engine

Clone the engine from the repo with `git clone https://github.com/thomasahle/sunfish.git`

## Creating a UCI engine binary
Since the engine is written in python, I have converted it into a binary using PyInstaller as follows:
`pyinstaller --onefile sunfish.py`

This resultant binary is found in "/sunfish/dist/sunfish/"

## Running UCI binary

This binary can be run in the terminal with success, with output shown below

```
(.venv) ➜  sunfish git:(master) ✗ ./dist/sunfish
uci
id name sunfish 2023
option name QS type spin default 40 min 0 max 300
option name QS_A type spin default 140 min 0 max 300
option name EVAL_ROUGHNESS type spin default 15 min 0 max 50
uciok
isready
readyok
go wtime 1000 btime 1000 winc 1000 binc 1000
info depth 1 time 0 nodes 2 nps 14488 score cp 0 lowerbound pv d2d4
info depth 1 time 0 nodes 4 nps 15266 score cp 46 upperbound
info depth 1 time 3 nodes 47 nps 16985 score cp 37 lowerbound pv g1f3
info depth 2 time 4 nodes 93 nps 20824 score cp 41 upperbound
bestmove g1f3
position startpos g1f3
isready
readyok
go wtime 1000 btime 1000 winc 1000 binc 1000
info depth 1 time 0 nodes 9 nps 32154 score cp 37 lowerbound pv g1f3 d7d5 d2d4
info depth 1 time 0 nodes 11 nps 22908 score cp 46 upperbound
info depth 2 time 2 nodes 70 nps 36441 score cp 41 upperbound
bestmove g1f3
```

## Issue
When trying to run sunfish via the en-croissant GUI, it times out. This happens no matter the length, 5/10/15 minutes. 
What is causing the engine to not work correctly with the GUI?

**Comment by franciscoBSalgueiro** [comment]
Does it work on the analysis board? There you can look at the engine logs. Also, is ./dist/sunfish a python file or a binary? You're running it in a virtual environment, but it has to work outside of it.

**Reply by C-Advait** [reply]
Hi Francisco, 

I've narrowed it down a little bit.
In the compiled/downloaded version of en-croissant, available [at this link for macOS](https://encroissant.org/download), neither the analysis board nor new-game are functional with the Sunfish engine.

However, when I build the code from source and run in dev mode, via the following procedure, both the new-game and analysis board work with the same options.

```
git clone https://encroissant.org/download
pnpm install
pnpm build
pnpm run dev
```

## Sunfish engine with dev en-croissant.
The following is a sample of the console output when playing new-game with Sunfish
```
    Finished dev [unoptimized + debuginfo] target(s) in 10.62s
[2024-05-14][17:26:01][INFO][en_croissant] Setting up application
[2024-05-14][17:26:01][INFO][en_croissant] Checking for required directories
[2024-05-14][17:26:01][INFO][en_croissant] Checking for required files
[2024-05-14][17:26:37][INFO][log@http://localhost:1420/node_modules/.vite/deps/tauri-plugin-log-api.js:19:20] React app started successfully
[2024-05-14][17:27:48][INFO][en_croissant::opening] Initializing openings table...
[2024-05-14][17:28:08][INFO][en_croissant::chess] Engine process finished: tab: f11b84d2black, engine: /Users/achauhan/Documents/projects/sunfish/dist/sunfish
```

## Compiling Sunfish
./dist/sunfish is a binary file.
It is a python repo, which can be converted into a binary via the following steps. I cannot attach the binary here because it is built for my computer specifications and I do not know the appropriate build target for other users.

```
git clone https://github.com/thomasahle/sunfish.git
cd directory
python3 -m pip install pyinstaller
pyinstaller --onefile sunfish.py
```
The binary should now be available at /dist/sunfish to use
The binary is tested outside a virtual environment, and the same functional behaviour is seen.

Thanks for following up so fast Francisco. I hope that the findings are helpful in narrowing down the problem.
My main goal for all this is to be able to see the "best_move" outputs and tweak them, while also playing through the gui.
If there is a better place for this discussion, please let me know.

Best,
Advait
